x/ccv/provider/keeper/punish_validator.go

```diff
@@ -31,8 +30,73 @@ func (k Keeper) JailValidator(ctx sdk.Context, providerAddr types.ProviderConsAd
 		k.stakingKeeper.Jail(ctx, providerAddr.ToSdkConsAddr())
 	}
 
-	// update jail time to end after double sign jail duration
+	// Jail the validator to trigger the unbonding of the validator
+	// (see cosmos/cosmos-sdk/blob/v0.45.16-ics-lsm/x/staking/keeper/val_state_change.go#L192).
 	k.slashingKeeper.JailUntil(ctx, providerAddr.ToSdkConsAddr(), evidencetypes.DoubleSignJailEndTime)
 
-	// TODO: add tombstoning back once we integrate the slashing
+	// Tombstone the validator so that we cannot slash the validator more than once
+	// (see cosmos/cosmos-sdk/blob/v0.45.16-ics-lsm/x/evidence/keeper/infraction.go#L81).
+	// Note that we cannot simply use the fact that a validator is jailed to avoid slashing more than once
+	// because then a validator could i) perform an equivocation, ii) get jailed (e.g., through downtime)
+	// and in such a case the validator would not get slashed when we call `SlashValidator`.
+	k.slashingKeeper.Tombstone(ctx, providerAddr.ToSdkConsAddr())
+}
+
+// ComputePowerToSlash computes the power to be slashed based on the tokens in non-matured `undelegations` and
+// `redelegations`, as well as the current `power` of the validator
+func (k Keeper) ComputePowerToSlash(ctx sdk.Context, validator stakingtypes.Validator, undelegations []stakingtypes.UnbondingDelegation,
+	redelegations []stakingtypes.Redelegation, power int64, powerReduction sdk.Int,
+) int64 {
+	// compute the total numbers of tokens currently being undelegated
+	undelegationsInTokens := sdk.NewInt(0)
+
+	cachedCtx, _ := ctx.CacheContext()
```

Comment:
Maybe instead of using this cached context here (or in addition to this), we could have wraps for SlashUnbondingDelegation and SlashRedelegation that just call those on cached contexts and call the wraps e.g. ComputeUnbondingDelegationPower - it's very subtle with the cached context and the slash factor of 1 here